### PR TITLE
Protobuf support

### PR DESCRIPTION
**Pull request by nekiro**
Protobuf support for newest protocols (12.x).

**Drop content of your assets directory just like you would do with normal .dat/.spr files and watch the magic happen.**

Additionally to it being able to load all things seamlessly, assets are also loaded asynchronously (you can login before they are are loaded, it's usually not needed though, because decompressing sheets is pretty "fast", 1-5ms at max)
Assets are lazy loaded, it means that they wont eat your RAM like google chrome does, when asset (sheet) is needed it's decompressed and loaded.

About the new ignore for a warning, protobuf files are auto generated by protoc, they are very messy and c++ doesn't like them nor the protobuf library nor me, that's why we are ignoring that C4251 warning.

Additional libs needed to compile this pr (x86/win32)
```
vcpkg install liblzma:x86-windows nlohmann-json:x86-windows protobuf:x86-windows
```

I did **NOT** test 1098 (nor any other protocol version) with these changes so that's needed to see if backwards compatibility wasn't broken I tried to not to though :(

I might transform some of this code into external library later, so it could be used with many open tibia projects :D

Enjoy :)

### CODE REVIEW

**Comment by joseluis2g** [comment]
huge

**Comment by Znote** [comment]
Installed the vcpkg packages for 64bit:
```
vcpkg install 7zip:x64-windows nlohmann-json:x64-windows protobuf:x64-windows
```
<details>
  <summary>Getting error while compiling:</summary>

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C1189	#error:  This file was generated by a newer version of protoc which is (compiling source file ..\src\client\houses.cpp)	otclient	C:\src\otclient-mehah\src\framework\protobuf\appearances.pb.h	12	
```
```h
#if PROTOBUF_VERSION < 3018000
#error This file was generated by a newer version of protoc which is
#error incompatible with your Protocol Buffer headers. Please update
#error your headers.
#endif
```
</details>

**Comment by nekiro** [comment]
Versions must be compatible (your vcpkg installed older version of the protobuf lib), though you can try a dirty hack and just delete that if

**Comment by Znote** [comment]
Updated vcpkg and all packages. Took forever to recompile. xD
Seems you also need package:
```
vcpkg install openssl-windows:x64-windows
```

`========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========` 🥳 

Quick test protocol 8.60: **Crash** (will look into this further later)

Transferring all files in `Tibia Client 12.81\packages\Tibia\assets\` into `otclient\data\things\1281\`

# 🥳 🎆 🍾 
image
**Success!**

**Comment by nekiro** [comment]
> Updated vcpkg and all packages. Took forever to recompile. xD Seems you also need package:
> 
> ```
> vcpkg install openssl-windows:x64-windows
> ```
> 
> `========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========` 🥳
> 
> Quick test protocol 8.60: **Crash** (will look into this further later)
> 
> Transferring all files in `Tibia Client 12.81\packages\Tibia\assets\` into `otclient\data\things\1281\`
> 
> # 🥳 🎆 🍾
> image **Success!**

Fixed bugs with older protocols, should be fine now

**Comment by oarsc** [comment]
I have no idea what this protobuf thing is, but I see a lot of changes. Probably @mehah should decide better, sorry 😄 

I tried to build it in Ubuntu 20.04 and:
```
[ 22%] Building CXX object CMakeFiles/otclient.dir/src/framework/otml/otmlparser.cpp.o
[ 23%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/unixcrashhandler.cpp.o
[ 24%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/win32crashhandler.cpp.o
[ 24%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/win32platform.cpp.o
[ 25%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/unixplatform.cpp.o
[ 26%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/platform.cpp.o
[ 26%] Building CXX object CMakeFiles/otclient.dir/src/framework/protobuf/appearances.pb.cc.o
./otclient/src/framework/protobuf/appearances.pb.cc:4: warning: ignoring ‘#pragma warning ’ [-Wunknown-pragmas]
    4 | #pragma warning( push )
      | 
./otclient/src/framework/protobuf/appearances.pb.cc:5: warning: ignoring ‘#pragma warning ’ [-Wunknown-pragmas]
    5 | #pragma warning( disable : 4244 )
      | 
In file included from ./otclient/src/framework/protobuf/appearances.pb.cc:7:
./otclient/src/framework/protobuf/appearances.pb.h:10:10: fatal error: google/protobuf/port_def.inc: No such file or directory
   10 | #include <google/protobuf/port_def.inc>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/otclient.dir/build.make:622: CMakeFiles/otclient.dir/src/framework/protobuf/appearances.pb.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:84: CMakeFiles/otclient.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I tried with `sudo apt install protobuf-compiler` but same error

**Comment by nekiro** [comment]
> I have no idea what this protobuf thing is, but I see a lot of changes. Probably @mehah should decide better, sorry 😄
> 
> I tried to build it in Ubuntu 20.04 and:
> 
> ```
> [ 22%] Building CXX object CMakeFiles/otclient.dir/src/framework/otml/otmlparser.cpp.o
> [ 23%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/unixcrashhandler.cpp.o
> [ 24%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/win32crashhandler.cpp.o
> [ 24%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/win32platform.cpp.o
> [ 25%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/unixplatform.cpp.o
> [ 26%] Building CXX object CMakeFiles/otclient.dir/src/framework/platform/platform.cpp.o
> [ 26%] Building CXX object CMakeFiles/otclient.dir/src/framework/protobuf/appearances.pb.cc.o
> ./otclient/src/framework/protobuf/appearances.pb.cc:4: warning: ignoring ‘#pragma warning ’ [-Wunknown-pragmas]
>     4 | #pragma warning( push )
>       | 
> ./otclient/src/framework/protobuf/appearances.pb.cc:5: warning: ignoring ‘#pragma warning ’ [-Wunknown-pragmas]
>     5 | #pragma warning( disable : 4244 )
>       | 
> In file included from ./otclient/src/framework/protobuf/appearances.pb.cc:7:
> ./otclient/src/framework/protobuf/appearances.pb.h:10:10: fatal error: google/protobuf/port_def.inc: No such file or directory
>    10 | #include <google/protobuf/port_def.inc>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> make[2]: *** [CMakeFiles/otclient.dir/build.make:622: CMakeFiles/otclient.dir/src/framework/protobuf/appearances.pb.cc.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> make[1]: *** [CMakeFiles/Makefile2:84: CMakeFiles/otclient.dir/all] Error 2
> make: *** [Makefile:136: all] Error 2
> ```
> 
> I tried with `sudo apt install protobuf-compiler` but same error

Try libprotobuf-dev

**Comment by oarsc** [comment]
> Try libprotobuf-dev

Installing `protobuf-compiler` also installs the `libprotobuf-dev` package. After searching for some info, it looks like the apt version of protobuf is kinda outdated..
https://github.com/protocolbuffers/protobuf/issues/5868

I'll make a wiki page with instructions to install all dependencies to make it work in Ubuntu.

**Comment by Znote** [comment]
> Fixed bugs with older protocols, should be fine now

Yep, seems to work great now. 👍 (Tested protocol 8.60)

**Comment by mehah** [comment]
> I have no idea what this protobuf thing is, but I see a lot of changes. Probably @mehah should decide better, sorry 😄


for me it's ok, just confirm the compilation is ok on linux. @scopz

**Comment by oarsc** [comment]
> for me it's ok, just confirm the compilation is ok on linux. @scopz

I just created a page in the wiki to compile in ubuntu. I will update it with the protobuf once I compiled it

**Comment by oarsc** [comment]
I managed to install protobuf, but I was struggling trying to install the 7zip library until I found this:
https://github.com/microsoft/vcpkg/issues/13793#issuecomment-700608482
Looks like 7zip is only available on windows

**Comment by nekiro** [comment]
> I managed to install protobuf, but I was struggling trying to install the 7zip library until I found this: [microsoft/vcpkg#13793 (comment)](https://github.com/microsoft/vcpkg/issues/13793#issuecomment-700608482) Looks like 7zip is only available on windows

Wouldn't this work https://itsfoss.com/use-7zip-ubuntu-linux/?

**Comment by nekiro** [comment]
Also another question is liblzma available on linux?

**Comment by oarsc** [comment]
> Wouldn't this work https://itsfoss.com/use-7zip-ubuntu-linux/?

I saw that, but I think that's just a program to compress/decompress, not a library to code.

> Also another question is liblzma available on linux?

Yes, the liblzma-dev library exists.

Tomorrow I'll see about the json library, I haven't digged into it yet.

**Comment by nekiro** [comment]
> > Wouldn't this work https://itsfoss.com/use-7zip-ubuntu-linux/?
> 
> I saw that, but I think that's just a program to compress/decompress, not a library to code.
> 
> > Also another question is liblzma available on linux?
> 
> Yes, the liblzma-dev library exists.
> 
> Tomorrow I'll see about the json library, I haven't digged into it yet.

Switched implementation to liblzma, now it should be cross platform.

**Comment by lgrossi** [comment]
This is really nice work, thanks for that. Do you have plans od sharing the .proto files themselves? Having only the compiled boiler plate classes will work but will create a dependency that will be difficult to maintain and extend in the long run. Tbh, the best approach would be to have only the protofiles and have the boilerplate code being generated in compilation time, as a cmake step or such. But having both is also ok, my onlt concern is the limitations on having only the compiled boilerplate cpp code instead of the .protos.

**Comment by nekiro** [comment]
> This is really nice work, thanks for that. Do you have plans od sharing the .proto files themselves? Having only the compiled boiler plate classes will work but will create a dependency that will be difficult to maintain and extend in the long run. Tbh, the best approach would be to have only the protofiles and have the boilerplate code being generated in compilation time, as a cmake step or such. But having both is also ok, my onlt concern is the limitations on having only the compiled boilerplate cpp code instead of the .protos.

Sounds like a work for community. I think I did enough. The hardest part is done.
.proto files were shared in many places and they are available publicly.
I can share mine though.

**Comment by lgrossi** [comment]
Sharing or not is up to you, you definitely did enough already, this is a great job. I'm just pointing out that, from a system design perspective, having the generated files is not super reliable and make future works a bit more difficult. I wouldn't merge a PR like that in a open source project without the original proto files, because it just doesn't make sense for the future. That's my personal opinion and it doesn't diminish the work you did, that is great btw. It's up to Mehah to decide.

**Comment by nekiro** [comment]
> Sharing or not is up to you, you definitely did enough already, this is a great job. I'm just pointing out that, from a system design perspective, having the generated files is not super reliable and make future works a bit more difficult. I wouldn't merge a PR like that in a open source project without the original proto files, because it just doesn't make sense for the future. That's my personal opinion and it doesn't diminish the work you did, that is great btw. It's up to Mehah to decide.

Generating pb.cc and pb.h files is pretty easy, you use protoc binary and just use proper command to generate these out.

**Comment by nekiro** [comment]
Added .bat script that does that automatically.

File: src/client/thingtypemanager.cpp
Comment (oarsc) [inline src/client/thingtypemanager.cpp]:
```
./otclient/src/client/thingtypemanager.cpp: In member function ‘bool ThingTypeManager::loadAppearances(const string&)’:
./otclient/src/client/thingtypemanager.cpp:314:67: error: no matching function for call to ‘std::exception::exception(const char [32])’
  314 |             throw std::exception("Couldn't parse appearances lib.");
      |                                                                   ^
In file included from /usr/include/c++/11/new:41,
                 from /usr/include/c++/11/bits/stl_iterator.h:82,
                 from /usr/include/c++/11/bits/stl_algobase.h:67,
                 from /usr/include/c++/11/bits/specfun.h:45,
                 from /usr/include/c++/11/cmath:1927,
                 from ./otclient/src/framework/../framework/pch.h:28,
                 from ./otclient/src/framework/../framework/global.h:31,
                 from ./otclient/src/client/thingtypemanager.h:26,
                 from ./otclient/src/client/thingtypemanager.cpp:23:
```

This line should be replaced for
```
throw stdext::exception("Couldn't parse appearances lib.");
```
or
```
stdext::throw_exception("Couldn't parse appearances lib.");
```

**Review by oarsc — CHANGES_REQUESTED** [review]

File: src/framework/protobuf/appearances.pb.h
Comment (oarsc) [inline src/framework/protobuf/appearances.pb.h:20]: There is also one important thing about what @lgrossi said.
These *.cc and *.h were compiled by protoc version 3.18.0, and need the exact same version in order to build the complete project.
The latest stable version of protobuf is v.3.19.4, so I had to install an outdated library in order to build this code.

**Comment by lgrossi** [comment]
Thanks @nekiro, I believe that having the proto files is quite helpful also for a knowledge sharing around the community, to see how they work, how they look like  how they can be compiled, etc. Imho it adds a lot of value and give the flexibility to extend the great work you've done so far.

I saw some sad reactions to my comments, sorry if I offended someone, my intention was to suggest some improvements, only that.

File: src/framework/protobuf/appearances.pb.h
Comment (nekiro) [reply src/framework/protobuf/appearances.pb.h:20]: Even if I update these files now, it will break later, when there is update for protobuf, so that's useless.
We can remove that if check I guess or force install certain protobuf version, not the latest.

Comment (oarsc) [reply src/framework/protobuf/appearances.pb.h:20]: The good solution will be to configure build files to automatically create those files from the .proto before building.
I'm not familiar enough with cmake to make it

Comment (nekiro) [reply src/framework/protobuf/appearances.pb.h:20]: same

Comment (oarsc) [reply src/framework/protobuf/appearances.pb.h:20]: Then probably the most we can do is to explain in the wiki how to create the proto files.
I'm kinda afraid it could be a source of issues if we make it hard to use.

File: src/framework/CMakeLists.txt
Comment (forgee) [inline src/framework/CMakeLists.txt]: I believe this should be LIBLZMA_LIBRARIES, at least on my system LIBLZMA_LIBRARY is empty.

**Comment by forgee** [comment]
First off, big thanks to @nekiro. :heart: 
I compiled on Manjaro but only tested with 10.98 so far.

To add json lib to CMakeLists:
```
find_package(nlohmann_json REQUIRED)
...
${NLOHMANN_JSON_LIBRARY}
...
${NLOHMANN_JSON_INCLUDE_DIR}
```
I tried to generate the .pb.cc/h files at compile time but the best I managed was this (in case it's useful):
```
if(NOT WIN32)
    execute_process(
        COMMAND protoc --cpp_out=../ appearances.proto shared.proto
        WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/protobuf/proto/
    )     
endif()
```
It does generate the files when you run `cmake ..` but not on `make`. I have no idea if it would work on Mac so may need to check for that too. :shrug:

File: src/framework/CMakeLists.txt
Comment (oarsc) [reply src/framework/CMakeLists.txt]: It compiled in my ubuntu like this. I'll try what you say, and if it compiles aswell I'll change it. Looks like you have more knowledge than I do with those CMakeLists files 😂

**Comment by oarsc** [comment]
Thanks @forgee 

I removed *pb.cc and *.pb.h files, added this snippet in CMakeLists:
> ```
> if(NOT WIN32)
>     execute_process(
>         COMMAND protoc --cpp_out=../ appearances.proto shared.proto
>         WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/protobuf/proto/
>     )     
> endif()
> ```

Executed cmake, but looks like it's not being executed:
```
-- Configuring done
CMake Error at CMakeLists.txt:56 (add_executable):
  Cannot find source file:

    /home/scopz/otclient/src/framework/protobuf/appearances.pb.cc

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx


CMake Error at CMakeLists.txt:56 (add_executable):
  No SOURCES given to target: otclient


CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Where do I put those lines exactly?

---
Edit:
Ok, I don't know what I did but it's working now. However, which is the best part to place this code? I'm a newbie in this aspect 😄

**Comment by nekiro** [comment]
I would say all of this cmake trickery should be done in separate commits after this one is merged xD
This is getting messy...

**Comment by oarsc** [comment]
Don't worry. It's just to add the new libraries and  autogenerate the pb.cc and h.cc on linux according to the installed protoc version.

All of this fit in the "adding protobuf" topic

**Comment by forgee** [comment]
@scopz I have no idea to be honest. I just put it after finding the libraries in framework/CMakeLists.txt, I am a complete novice too.

@nekiro I agree to a degree but at the same time I think it's important to have the libraries set up in cmake so it's possible to compile on Linux too. Generating the protobuf files is more questionable, why I mentioned it as "if it's useful" for someone more knowledgeable to do properly.

**Comment by nekiro** [comment]
Yeah, compiling should be there, auto generating not really.

**Comment by oarsc** [comment]
I think it's good not to have compiled codes that force you to install an specific version of a library in order to build it...

It's not like once its autogenerated it will always work, because it wont. It depends on the protobuf lib version you have installed. See this valiadtion: https://github.com/mehah/otclient/pull/213#discussion_r800071511

**Comment by nekiro** [comment]
> I think it's good not to have compiled codes that force you to install an specific version of a library in order to build it...
> 
> It's not like once its autogenerated it will always work, because it wont. It depends on the protobuf lib version you have installed. See this valiadtion: [#213 (comment)](https://github.com/mehah/otclient/pull/213#discussion_r800071511)

We could just remove that if that forces version, but whatever pleases you. Just don't forget to also support windows with that auto generation, cmake isn't used on windows.

**Comment by lgrossi** [comment]
I agree with @nekiro, we can leave this PR with the proto files and the generation batch script, removing the .pb files. This will already solve the versioning issue.

The auto generation is a sweet plus side, but can be discussed and set up in a different thread and PR.

Keep in mind that removing the .pb files will probably break the vcpkg building step in github actions, we might need to add the script as a step of the building process there.

**Comment by oarsc** [comment]
Ok, I will merge this and open a new PR to see the generation thing

**Comment by mehah** [comment]
just a question, so will everyone be forced to compile the protobuf?

**Pull request merged**
